File: macos/Sources/AndroMacKit/FileNames.swift
import Foundation

/// File names that arrive over the wire are untrusted (docs/PROTOCOL.md §5, receiver rules).
/// Pure functions, so they can be tested without a socket.
public enum FileNames {

    /// The file system limit on one path component.
    public static let maxBytes = 255

    /// Last path component only, no control characters, no leading dots or whitespace, at most
    /// 255 bytes of UTF-8 (the stem is cut, the extension is kept). `file` when nothing is left.
    public static func sanitize(_ raw: String) -> String {
        var name = raw.split(whereSeparator: { $0 == "/" || $0 == "\\" }).last.map(String.init) ?? ""
        name.removeAll { c in c.unicodeScalars.contains { $0.value < 0x20 || $0.value == 0x7F } }
        name = String(name.drop { $0 == "." || $0.isWhitespace })
        if name.utf8.count > maxBytes {
            let (stem, ext) = splitExtension(name)
            let cutStem = cut(stem, to: maxBytes - ext.utf8.count)
            name = cutStem.isEmpty ? cut(name, to: maxBytes) : cutStem + ext
        }
        return name.isEmpty ? "file" : name
    }

    /// `photo.jpg` → `photo (2).jpg`, `photo (3).jpg`, … until `exists` says no.
    public static func unique(_ name: String, exists: (String) -> Bool) -> String {
        guard exists(name) else { return name }
        let (stem, ext) = splitExtension(name)
        // Leave room for the suffix so a name that already used the full 255 bytes still fits.
        let base = cut(stem, to: maxBytes - ext.utf8.count - 8)
        for n in 2...9999 {
            let candidate = "\(base) (\(n))\(ext)"
            if !exists(candidate) { return candidate }
        }
        return "\(base) (\(UUID().uuidString.prefix(8)))\(ext)"
    }

    /// `("photo", ".jpg")`; a name without a dot has an empty extension.
    static func splitExtension(_ name: String) -> (stem: String, ext: String) {
        guard let dot = name.lastIndex(of: "."), dot != name.startIndex else { return (name, "") }
        return (String(name[..<dot]), String(name[dot...]))
    }

    /// Truncates on a character boundary so no scalar is split in half.
    private static func cut(_ text: String, to bytes: Int) -> String {
        var out = ""
        for c in text {
            if out.utf8.count + c.utf8.count > bytes { break }
            out.append(c)
        }
        return out
    }
}

/// The numbers of PROTOCOL §5: 512 KiB chunks, files up to 4 GiB.
public enum FileTransferMath {

    public static let chunkSize = 524_288
    public static let maxFileSize: Int64 = 4 << 30

    /// How many `file_chunk` messages a file of `size` bytes takes; `0` for an empty file.
    public static func chunkCount(size: Int64) -> Int {
        Int((size + Int64(chunkSize) - 1) / Int64(chunkSize))
    }
}

File: android/app/src/main/kotlin/dev/andromac/core/FileNames.kt
package dev.andromac.core

/**
 * Receiver-side file name rules from PROTOCOL §5, applied to whatever the sender claims.
 * Pure Kotlin so the vectors module can unit-test it on a plain JVM.
 */
object FileNames {

    private const val MAX_BYTES = 255
    private const val FALLBACK = "file"

    /**
     * Last path component only, no control or invisible format characters, no leading dots or
     * whitespace, at most 255 bytes of UTF-8 (the stem is cut, the extension kept), `file` when
     * nothing is left. Format characters (category Cf) are the bidi overrides and zero-width
     * ones, which can make `x\u202Egnp.apk` read as `xkpa.png`.
     */
    fun sanitize(name: String): String {
        var s = name.substringAfterLast('/').substringAfterLast('\\')
        s = buildString {
            s.codePoints().forEach { cp ->
                if (cp >= 0x20 && cp != 0x7F && Character.getType(cp) != Character.FORMAT.toInt()) appendCodePoint(cp)
            }
        }
        s = s.trimStart { it == '.' || it.isWhitespace() }.trimEnd()
        if (s.isEmpty()) return FALLBACK
        if (s.toByteArray().size <= MAX_BYTES) return s

        val (stem, ext) = split(s)
        // An absurdly long "extension" is not one; cut the whole name instead.
        val keepExt = ext.toByteArray().size <= 32
        val budget = MAX_BYTES - if (keepExt) ext.toByteArray().size else 0
        val cut = truncateUtf8(if (keepExt) stem else s, budget).trimEnd { it == '.' || it.isWhitespace() }
        val out = if (keepExt) cut + ext else cut
        return out.ifEmpty { FALLBACK }
    }

    /** `photo.jpg` -> `photo (2).jpg` while [exists] says the name is taken. Counts from 2, like Finder. */
    fun unique(name: String, exists: (String) -> Boolean): String {
        if (!exists(name)) return name
        val (stem, ext) = split(name)
        var n = 2
        while (true) {
            val candidate = "$stem ($n)$ext"
            if (!exists(candidate)) return candidate
            n++
        }
    }

    /** ("photo", ".jpg"); a name with no dot, or only a leading one, has no extension. */
    private fun split(name: String): Pair<String, String> {
        val dot = name.lastIndexOf('.')
        return if (dot <= 0) name to "" else name.substring(0, dot) to name.substring(dot)
    }

    /** Longest prefix that fits [bytes] of UTF-8 without cutting a code point. */
    private fun truncateUtf8(s: String, bytes: Int): String {
        var used = 0
        val out = StringBuilder()
        var i = 0
        while (i < s.length) {
            val cp = s.codePointAt(i)
            val len = String(Character.toChars(cp)).toByteArray().size
            if (used + len > bytes) break
            out.appendCodePoint(cp)
            used += len
            i += Character.charCount(cp)
        }
        return out.toString()
    }
}
